### DISCUSSION TITLE
PCAPs Not getting cleaned up filling /nsm

### DISCUSSION THREAD

**Discussion by pgatty** [Security-Onion-Solutions/securityonion#12184] [2.4] [answered]
### Version

2.4.0

### Installation Method

Security Onion ISO image

### Description

other (please provide detail below)

### Installation Type

Standalone

### Location

on-prem with Internet access

### Hardware Specs

Meets minimum requirements

### CPU

4

### RAM

16GB

### Storage for /

163GB

### Storage for /nsm

327GB

### Network Traffic Collection

span port

### Network Traffic Speeds

1Gbps to 10Gbps

### Status

Yes, all services on all nodes are running OK

### Salt Status

No, there are no failures

### Logs

No, there are no additional clues

### Detail

I've been running Security Onion 2.4 for several months now and everything has been fine until a few weeks ago when whatever job manages the PCAP space stopped cleaning up old files. Prior to this issue the free space on /nsm partition would stay at 60%. Now PCAPs will completely fill up the disk space necessitating a manual purge of PCAPs. I changed the diskfreepercentage from 10 to 25 under Administration –> Configuration –> pcap –> config –> diskfreepercentage and synced the grid, but the free disk space will drop all the way to 0%. I've looked through the stenographer log and don't see any errors. I've tried looking through some of the other logs in /opt/so/log but nothing stands out as a problem. Currently /nsm is at 88% used,

Filesystem               Size  Used Avail Use% Mounted on
devtmpfs                 4.0M     0  4.0M   0% /dev
tmpfs                    7.6G   32K  7.6G   1% /dev/shm
tmpfs                    3.1G  323M  2.8G  11% /run
/dev/mapper/system-root  163G   26G  137G  16% /
/dev/sda2                507M  368M  140M  73% /boot
/dev/sda1               1022M  6.2M 1016M   1% /boot/efi
/dev/mapper/system-tmp   2.0G  149M  1.8G   8% /tmp
/dev/mapper/system-nsm   327G  286G   41G  88% /nsm
tmpfs                    1.6G     0  1.6G   0% /run/user/939
tmpfs                    1.6G     0  1.6G   0% /run/user/0
tmpfs                    1.6G     0  1.6G   0% /run/user/1000

Here's the output of sudo du -csh * from /nsm
7.2G    backup
6.2G    docker-registry
2.8G    elastic-fleet
114G    elasticsearch
0       import
1.1G    influxdb
600K    kratos
4.0K    logstash
270M    mysql
140G    pcap
2.7G    pcapindex
0       pcapout
0       pcaptmp
216M    redis
8.6G    repo
65M     rules
0       soc
28M     strelka
284K    suricata
496M    zeek
283G    total

### Guidelines

- [X] I have read the above statement and can confirm my post is relevant to Security Onion 2.4.

**Comment by petiepooo** [comment]
What is the latest contents of /opt/so/logs/so-sensor-clean, and does its timestamp match the current date/time?  Cron should run that every minute for zeek, strelka, and suricata logs, but pcap (stenographer) and elasticsearch (curator) should be cleaning themselves separately.

**Reply by pgatty** [reply]
I've checked the contents of /opt/so/logs/sensor_clean.log and see the jobs cleaning up strelka, and suricata files then following repeated continuously:

Tue Jan 16 02:19:16 AM UTC 2024 - No old Zeek logs available to clean up in /nsm/zeek/logs/
Tue Jan 16 02:19:16 AM UTC 2024 - No old files available to clean up in /nsm/strelka/processed
Tue Jan 16 02:19:16 AM UTC 2024 - No old files available to clean up in /nsm/suricata
Tue Jan 16 02:19:16 AM UTC 2024 - No old files available to clean up in /nsm/pcapout

The odd thing is I only see entries in the sensor_clean.log for a very narrow time slice on the 16th between 02:19:02 AM UTC and 02:52:53 AM UTC. The logs for the 14th and 15th are empty.

**Reply by petiepooo** [reply]
I believe that's normal, as the so-sensor-clean script is silent if the disk does not need cleaning.  The time range you see in those output lines would be during the time it is trying to free up space, right?  The output just shows there's nothing to delete among the files it manages, as the space is used by something else, like pcap.

I can't contribute further, sorry.  Perhaps the devs can add more.

**Comment by pgatty** [comment]
@petiepooo - Thanks for taking the time to respond and trying to help. 

So I ended up rebooting my Security Onion instance this morning to see if that would fix anything and within a few minutes after the reboot, something kicked in and removed 85GB of PCAP files. So the problem is clearly with PCAP clean up and it appears that something happens that causes the stenographer clean up job to fail. Looking through the stenographer log at the time of the reboot, I can see it clean up all the pcap index files in /nsm/pcapindex for the PCAPs I manually deleted, but I don't see any log entries for the deletion of the 85GB of PCAP files in /nsm/pcap.

I will continue to monitor the situation, If this happens again, is there anything I can check within stenographer to see if I can locate the problem or check the health of stenographer beyond running so-status?

**Reply by pgatty** [reply]
This continues to be a problem. Free space on /nsm dropped to 8% and there's nothing in the stenographer log that indicates any errors. Rebooting the server clears all the files out /nsm/pcap.  Here's what diskfreepercentage is set to. I'm assuming this means that stenographer would remove old pcaps when the free space % on /nsm drops below 25%. Is this correct?

![image](https://github.com/Security-Onion-Solutions/securityonion/assets/8431123/a01e4291-8fee-4a76-bf32-9718db3eb24c)

**Reply by petiepooo** [reply]
The next time this happens to you, here's what I would recommend:
1. run `sudo docker logs so-steno` to see if there's anything outside of redirected stdio to indicate a fault
2. run `sudo docker top so-steno` to see what processes are running in the container
3. run `sudo so-restart steno` to restart the container (this may be an easier way to recover rather than rebooting)
4. check in a minute to see if space in /nsm is clearing again, and try something else (like a reboot) if not
5. run `sudo docker top so-steno` and compare the output to what you saw above

You should see `stenographer` and `stenotype` as well as `runuser` and a couple of mediating shells.  Stenographer itself is the process that's supposed to be watching disk space.  Stenotype is its child process and does the actual read/index/write of pcaps.  If restarting so-steno worked, next time you might try killing just stenotype to see if stenographer relaunches it as it should.  

You mention traffic speeds vary from 1-10Gbps.  Did you verify that via InfluxDB screens?  How many days of PCAP retention does InfluxDB show?  Is the failure triggered around the same time as a spike in monitored traffic?  327GB disk is not much space for your /nsm partition if you're really trying to ingest that much traffic, but neither is 4CPU or 16GB RAM.  At 2Gbps ingest, I'm seeing stenotype itself use over 2GB RAM and only getting a couple days of PCAP retention with over 35TB disk.  If you're running on a VM, you might try doubling or tripling everything to see if that helps.

**Reply by pgatty** [reply]
The /nsm partition has nearly filled up several times now. Each time I've run through the steps you outlined above.

1. run sudo docker logs so-steno - I just see these unrelated messages 
```
Skipping CA state generation, ca_cert.pem already exists
Skipping generation of 'client_127.0.0.1_client' key/cert, client_127.0.0.1_client_cert.pem already exists
Skipping generation of 'server_127.0.0.1' key/cert, server_127.0.0.1_cert.pem already exists
```
2. run sudo docker top so-steno - nothing interesting here, just the same output each time with the expected processes. The output is the same both before and after I restart the steno container to clear out the logs.
```UID                 PID                 PPID                C                   STIME               TTY                 TIME                CMD
root                2218630             2218582             0                   Jan22               ?                   00:00:00            /bin/bash /usr/local/sbin/so-steno.sh
root                2218911             2218630             0                   Jan22               ?                   00:00:00            runuser -l stenographer -c /usr/bin/stenographer --syslog=false >> /var/log/stenographer/stenographer.log 2>&1
stenogr+            2218931             2218911             0                   Jan22               ?                   00:00:00            -bash -c /usr/bin/stenographer --syslog=false >> /var/log/stenographer/stenographer.log 2>&1
stenogr+            2218958             2218931             0                   Jan22               ?                   00:07:34            /usr/bin/stenographer --syslog=false
stenogr+            2219098             2218958             0                   Jan22               ?                   00:02:16            /usr/bin/stenotype -v --blocks=2048 --preallocate_file_mb=4096 --aiops=128 --uid=stenographer --gid=stenographer --threads=1 --dir=/tmp/stenographer643328816 --iface=bond0
```
3. run sudo so-restart steno - The good news is restarting the container clears out the PCAPs

Regrading your questions about traffic speeds. My security onion instance is tapping a 10g port and has a separate, dedicated 10g interface for this, but the average bandwidth is only around 3 - 4 Mbps with spikes up to 800Mbps. I was running the 2.3 version of Security Onion with the exact same specs and it worked fine. Yes 327GB for /nsm isn't a lot, but it's enough retention for my needs.  The clean up process for PCAPs is supposed to be based on fee disk percentage, go the size of /nsm really only impacts retention. The bottom line here is that the clean job for PCAP's is broken. I'm going to update to 2.4.40 in a last ditch effort to fix the problem.

**Reply by petiepooo** [reply]
That it worked in 2.3 is promising, but if you have time to do a little more debugging, here's how to get logging of the stenographer process working (currently only stenotype is logging to file).

exec into the so-steno container:
> docker exec -ti so-steno bash

edit the launch script:
> vi /usr/local/bin/so-steno.sh

add a "-v 2" option before the "--syslog=false" option on the last line, then save and exit

prevent the container from stopping after next step:
> killall -STOP runuser

exit stenographer and stenotype:
> killall stenographer

wait until stenographer dies (repeat until only runuser, bash, and ps are left, one bash should show as defunct):
> ps -ef

relaunch stenographer with verbose logging:
> /usr/local/bin/so-steno.sh &

disown the children and exit back to host:
> disown
exit

once you are out of the container, follow stenographer's output with:
> tail -f /opt/so/log/stenographer/stenographer.log

You should see lines like "Thread 0 disk space is sufficient ..." every 15 seconds.  You can exit the tail with Ctrl-C and go about your business.  When it stops working, see what the logfile says; probably around the point where those Thread 0 messages stop.

That said, if you're on 2.4.0, not 10, 20, or 30, definitely update.. there were lots of bugs squashed since then.  I didn't see any specifically for this, but who knows.

**Reply by petiepooo** [reply]
See https://github.com/Security-Onion-Solutions/securityonion-image/pull/509

**Reply by pgatty** [reply]
I tried getting into the so-steno container and it doesn't work. If I run `docker exec -ti so-steno bash` as my normal user I get a permission denied message. If I run with sudo it works, but the file `/usr/local/bin/so-steno.sh` doesn't exist. Apologies I'm a Docker newb. Thanks again for your efforts to help be debug this.

**Reply by petiepooo** [reply]
Yes, docker would need to be run as root unless your user is a member of the docker group.  

My mistake; the file to edit is at /usr/local/sbin/so-steno.sh.

**Reply by petiepooo** [reply]
And you can use nano instead of vi if you prefer, of course.

**Reply by pgatty** [reply]
I managed to edit the file, but the killall command doesn't exist in the container. Should I just kill all the stenographer processes running as root and stenographer using their PID's i.e. kill -9 PID then relaunch stenographer with verbose logging via `/usr/local/bin/so-steno.sh &`

**Comment by petiepooo** [comment]
Yes, but not with the -9.  Find the pid of runuser and run `kill -STOP <pid>`.  Then kill stenographer with just `kill <pid>`.

FWIW, the -9 option for kill tells the kernel to remove the process from its process table.  It's the "nuke from high orbit" option, and I don't recommend it except as a last ditch effort.  It's more polite to send a normal -TERM (-15, the default) signal to the process first and "ask" it to clean up and terminate itself rather than yank the rug out from under it with a -KILL (-9)...  

The -STOP (-19) to runuser, BTW, tells the kernel to stop scheduling the process, which is why it hangs around in the ps output.  If it were to run again, say by sending -CONT (-18), it would see its child the (defunct) shell has exited and would exit itself, eventually stopping the entire container.  We don't want that in this case as we want to manually restart the processes in the existing container.

I was testing in the 2.3.280 so-steno image.  Killall is there but must not be in your newer 2.4 image... oh well.  Thanks for sticking with the debugging effort.

**Comment by pgatty** [comment, marked as answer]
I thought I'd post an update to this. I was never able to figure out what was causing the problem and ended up reinstalling Security Onion (2.4.50) rather than have /nsm fill up every 3 - 4 days. It's been running for over a week now and PCAP cleanup is working as expected. Sometimes it's just easier to nuke the site from orbit ;). I wanted to thank @petiepooo for all his help.
